controller/program/src/instructions/swap_exact_in.rs

```diff
@@ -0,0 +1,394 @@
+    let lst_state_list = *ix_prefix.lst_state_list();
+    let inp_lst_token_program = *ix_prefix.inp_lst_token_program();
+    let out_lst_token_program = *ix_prefix.out_lst_token_program();
+    let inp_pool_reserves = *ix_prefix.inp_pool_reserves();
+    let out_pool_reserves = *ix_prefix.out_pool_reserves();
+
+    let list = LstStatePackedList::of_acc_data(abr.get(lst_state_list).data())
+        .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidLstStateListData))?;
+
+    let (inp_lst_state, expected_inp_reserves) = get_lst_state_data(
+        abr,
+        &list,
+        args.inp_lst_index as usize,
+        inp_lst_token_program,
+    )?;
+    let (out_lst_state, expected_out_reserves) = get_lst_state_data(
+        abr,
+        &list,
+        args.out_lst_index as usize,
+        out_lst_token_program,
+    )?;
+
+    if inp_lst_state.mint == out_lst_state.mint {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::SwapSameLst).into());
+    }
+
+    let expected_protocol_fee_accumulator = create_raw_protocol_fee_accumulator_addr(
+        abr.get(out_lst_token_program).key(),
+        &out_lst_state.mint,
+        &out_lst_state.protocol_fee_accumulator_bump,
+    )
+    .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidReserves))?;
+
+    // Verify incoming accounts
+    let expected_pks = NewSwapExactInIxPreAccsBuilder::start()
+        .with_lst_state_list(&LST_STATE_LIST_ID)
+        .with_pool_state(&POOL_STATE_ID)
+        .with_protocol_fee_accumulator(&expected_protocol_fee_accumulator)
+        .with_inp_pool_reserves(&expected_inp_reserves)
+        .with_out_pool_reserves(&expected_out_reserves)
+        .with_inp_lst_mint(&inp_lst_state.mint)
+        .with_out_lst_mint(&out_lst_state.mint)
+        .with_inp_lst_token_program(abr.get(*ix_prefix.inp_lst_mint()).owner())
+        .with_out_lst_token_program(abr.get(*ix_prefix.out_lst_mint()).owner())
+        // NOTE: For the following accounts, it's okay to use the same ones passed by the user since the CPIs would fail if they're not as expected.
+        // User can't pass the `inp_lst_reserves` as `inp_lst_acc` because we're also not doing `invoke_signed` for the `inp_lst` transfer.
+        .with_inp_lst_acc(abr.get(*ix_prefix.inp_lst_acc()).key())
+        .with_out_lst_acc(abr.get(*ix_prefix.out_lst_acc()).key())
+        .with_signer(abr.get(*ix_prefix.signer()).key())
+        .build();
+
+    verify_pks(abr, &ix_prefix.0, &expected_pks.0)?;
+
+    if U8Bool(&inp_lst_state.is_input_disabled).is_true() {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::LstInputDisabled).into());
+    }
+
+    // safety: account data is 8-byte aligned
+    let pool = unsafe { PoolState::of_acc_data(abr.get(pool_state).data()) }
+        .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidPoolStateData))?;
+    verify_not_rebalancing_and_not_disabled(pool)?;
+
+    let (inp_calc_all, suf) = suf
+        .split_at_checked(args.inp_lst_value_calc_accs as usize)
```

Comment:
nit: `u8` impls `Into<usize>`. Old advice about treating `as` like the devil applies
```suggestion
        .split_at_checked(args.inp_lst_value_calc_accs.into())
```